Report field name of incorrect JSON types

guss77 opened the pull request:
With the current `fromJson()` implementation, if the user sends an unexpected type in a JSON field (e.g. send a number for a string field), the Vert.x `JsonObject` implementation will cause a `ClassCastException` to be thrown.

While that exception contains information about the type that failed to convert (both the actual type and the expected type), it doesn't have information about the name of the field that failed, making it harder than necessary to figure out the problem.

This change adds a catch clause to each field getter and handles a `ClassCastException` by throwing a new exception with a message that encodes the field name where the problem occurred as well as a hopefully more friendly format for the type names (without all that Java package stuff).

guss77 commented:
I think a more useful approach would be to not throw a new `ClassCastException` with a weird message, but create a new exception class that encodes the field name and expected type (it can extend 'ClassCastException` for backward compatibility) then a Vert.x failure handler can detect it and format whatever message the developer wants to share with the end user.

Maybe something like this:

```
throw new UnexpectedJsonValueType("field", "field-type", e);
```

The new exception to type would have to be added to vertx-jooq-shared, of course.

Does that sound like a good idea?

jklingsporn commented:
Thanks again for your contribution. I really appreciate it!
I prefer the second option over a random ClassCastException that can be thrown at any point in the application. On the other hand I prefer reusing existing exceptions over creating new ones but I don't think there is one in `java.lang` (`IllegalArgumentException` comes closest but does not reflect the issue completely).

guss77 commented:
Thanks for the feedback - I've moved the error into its own type as discussed. 

I hope the package is OK - please let me know any comments you have, I want to make this 100% great as per your usual standards :-)

There's also a serious bug with the previous patch, which I fixed by removing the relevant code completely - I no longer try to parse the `ClassScastException` message.

mahieddine commented:
Hi, 
While the new PR for vertx-jooq has been accepted, the PR which includes the new class UnexpectedJsonValueType has not been merged yet into vertx-jooq-shared package so when the POJOs are generated, there is an error which says the class **UnexpectedJsonValueType** is not found.

guss77 commented:
@mahieddine - to which code branch are you referring? The branch `release4.2.0` includes the entire PR, including the new exception type, [as can be seen here](https://github.com/jklingsporn/vertx-jooq/blob/release4.2.0/vertx-jooq-shared/src/main/java/io/github/jklingsporn/vertx/jooq/shared/UnexpectedJsonValueType.java)

guss77 commented:
Works fine in current stable release 4.2.0 from Maven Central

mahieddine commented:
Hi @guss77 my bad i deleted ~/.m2 dir and compiled again and it worked, guess i had something nasty in my local repos
